File: cpp/gtest/Parrot.cpp
#include <stdexcept>
#include <algorithm>

#include "Parrot.h"

using namespace std;

Parrot::Parrot(ParrotType parrotType, int numberOfCoconuts, double voltage, bool isNailed)
: parrotType(parrotType), numberOfCoconuts(numberOfCoconuts), voltage(voltage), isNailed(isNailed) {}

double Parrot::getBaseSpeed() {
    return 12.0;
}

double Parrot::getBaseSpeed(double current_voltage) {
    return min(24.0, current_voltage * getBaseSpeed());
}


double EuropeanParrot::getSpeed()
{
    return getBaseSpeed();
}


AfricanParrot::AfricanParrot(int numberOfCoconuts)
    : Parrot{DERIVED, numberOfCoconuts, 0., false}
    , numberOfCoconuts{ numberOfCoconuts }
{}

double AfricanParrot::getSpeed()
{
    return max(0.0, getBaseSpeed() - getLoadFactor() * numberOfCoconuts);
}

double AfricanParrot::getLoadFactor() {
    return 9.0;
}

double NorwegianBlueParrot::getSpeed()
{
    return (isNailed) ? 0 : getBaseSpeed(voltage);
}

File: JavaScript/src/parrot.js
export const PARROT_TYPES = {
    EUROPEAN:       'EUROPEAN',
    AFRICAN:        'AFRICAN',
    NORWEGIAN_BLUE: 'NORWEGIAN_BLUE',
};

class UnknownParrotError extends Error {
    constructor(type) {
        super(`Unknown parrot: ${type}`);
    }
}

export class Parrot {

    constructor(type, numberOfCoconuts, voltage, isNailed) {
        this.type = type;
        this.numberOfCoconuts = numberOfCoconuts;
        this.voltage = voltage;
        this.isNailed = isNailed;
    }

    getSpeed() {
        switch (this.type) {
            case PARROT_TYPES.EUROPEAN:
                const europeanParrot = new EuropeanParrot();
                return europeanParrot.getSpeed();
            case PARROT_TYPES.AFRICAN:
                const africanParrot = new AfricanParrot(this.numberOfCoconuts);
                return africanParrot.getSpeed();
            case PARROT_TYPES.NORWEGIAN_BLUE:
                const norwegianBlueParrot = new NorwegianBlueParrot(this.voltage, this.isNailed);
                return norwegianBlueParrot.getSpeed();
        }
        this.handleUnknownParrot();
    }

    handleUnknownParrot() {
        throw new UnknownParrotError(this.type);
    }

    getBaseSpeed() {
        const baseSpeed = 12;
        return baseSpeed;
    }
}

class EuropeanParrot extends Parrot {
    constructor() {
        super(PARROT_TYPES.EUROPEAN, undefined, undefined, undefined);
    }

    getSpeed() {
        return super.getBaseSpeed();
    }
}

class AfricanParrot extends Parrot {

    constructor(numberOfCoconuts) {
        super(PARROT_TYPES.AFRICAN, numberOfCoconuts, undefined, undefined);
        this.loadFactor = 9;
        this.minimalSpeed = 0;
    }

    getSpeed() {
        return Math.max(0, this.getBaseSpeed() - this.loadFactor * this.numberOfCoconuts);
    }
}

class NorwegianBlueParrot extends Parrot {
    constructor(voltage, isNailed) {
        super(PARROT_TYPES.NORWEGIAN_BLUE, undefined, voltage, isNailed);
    }

    getSpeed() {
        return (this.isNailed) ? 0 : this.getBaseSpeedWithVoltage(this.voltage);
    }

    getBaseSpeedWithVoltage(voltage) {
        const minimalSpeedTreshold = 24;
        const actualSpeed = voltage * this.getBaseSpeed();
        const baseSpeed = Math.min(minimalSpeedTreshold, actualSpeed);
        return baseSpeed;
    }
}

